pkg/manifest/objects.go

```diff
@@ -0,0 +1,343 @@
+	err := decoder.Decode(out)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding object: %v", err)
+	}
+	return NewObject(out, nil, yaml), nil
+}
+
+// UnstructuredObject exposes the raw object, primarily for testing
+func (o *Object) UnstructuredObject() *unstructured.Unstructured {
+	return o.object
+}
+
+// GroupKind returns the GroupKind for o.
+func (o *Object) GroupKind() schema.GroupKind {
+	return o.object.GroupVersionKind().GroupKind()
+}
+
+// GroupVersionKind returns the GroupVersionKind for o.
+func (o *Object) GroupVersionKind() schema.GroupVersionKind {
+	return o.object.GroupVersionKind()
+}
+
+// Hash returns a unique hash for o.
+func (o *Object) Hash() string {
+	return Hash(o.Kind, o.Namespace, o.Name)
+}
+
+// HashNameKind returns a hash for o based on name and kind only.
+func (o *Object) HashNameKind() string {
+	return HashNameKind(o.Kind, o.Name)
+}
+
+// JSON returns a JSON representation of o, using an internal cache.
+func (o *Object) JSON() ([]byte, error) {
+	if o.json != nil {
+		return o.json, nil
+	}
+
+	b, err := o.object.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = b
+	return b, nil
+}
+
+// YAML returns a YAML representation of o, using an internal cache.
+func (o *Object) YAML() ([]byte, error) {
+	if o.yaml != nil {
+		return o.yaml, nil
+	}
+	// TODO: there has to be a better way.
+	oj, err := o.JSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = oj
+	y, err := yaml.JSONToYAML(oj)
+	if err != nil {
+		return nil, err
+	}
+	o.yaml = y
+	return y, nil
+}
+
+// YAMLDebugString returns a YAML representation of o, or an error string if the object cannot be rendered to YAML.
+func (o *Object) YAMLDebugString() string {
+	y, err := o.YAML()
+	if err != nil {
+		return fmt.Sprint(err)
+	}
+	return string(y)
+}
+
+// AddLabels adds labels to o.
+func (o *Object) AddLabels(labels map[string]string) {
+	merged := make(map[string]string)
+	for k, v := range o.object.GetLabels() {
+		merged[k] = v
+	}
+
+	for k, v := range labels {
+		merged[k] = v
+	}
+
+	o.object.SetLabels(merged)
+	// Invalidate cached json
+	o.json = nil
+	o.yaml = nil
+}
+
+// Objects holds a collection of objects, so that we can filter / sequence them
+type Objects []*Object
+
+// ParseObjectsFromYAMLManifest returns an Objects represetation of manifest.
+func ParseObjectsFromYAMLManifest(manifest string) (Objects, error) {
+	var b bytes.Buffer
+
+	var yamls []string
+	scanner := bufio.NewScanner(strings.NewReader(manifest))
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "---" {
+			// yaml separator
+			yamls = append(yamls, b.String())
+			b.Reset()
+		} else {
+			if _, err := b.WriteString(line); err != nil {
+				return nil, err
+			}
+			if _, err := b.WriteString("\n"); err != nil {
+				return nil, err
+			}
+		}
+	}
+	yamls = append(yamls, b.String())
+
+	var objects Objects
+
+	for _, yaml := range yamls {
+		// We need this so we don't error on a file that is commented out
```

Comment:
Couldn't we use `ParseYAMLToObject` here and also dealing with commented manifest snippets by checking the parsed object for `Kind`?

```
o, err := ParseYAMLToObject([]byte(yaml))
if err != nil {
	log.Infof("error decoding object: %s\n%s\n", err, yaml)
	continue
}

if o.GroupKind().Group == "" {
	continue
}

objects = append(objects, o)
```

Reply:
Yes, that's a great suggestion, thanks.